tiledb/sm/query/update_value.cc

```diff
@@ -0,0 +1,159 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Implements the UpdateValue class.
+ */
+
+#include "tiledb/sm/query/update_value.h"
+#include "tiledb/common/logger.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+UpdateValue::UpdateValue(
+    std::string field_name,
+    const void* update_value,
+    const uint64_t update_value_size)
+    : field_name_(field_name)
+    , update_value_data_(update_value_size)
+    , update_value_view_(
+          (update_value != nullptr && update_value_size == 0 ?
+               (void*)"" :
+               update_value_data_.data()),
+          update_value_data_.size()) {
+  if (update_value_view_.size() != 0) {
+    memcpy(update_value_data_.data(), update_value, update_value_size);
+  }
+}
+
+UpdateValue::UpdateValue(const UpdateValue& rhs)
+    : field_name_(rhs.field_name_)
+    , update_value_data_(rhs.update_value_data_)
+    , update_value_view_(
+          (rhs.update_value_view_.content() != nullptr &&
+                   rhs.update_value_view_.size() == 0 ?
+               (void*)"" :
+               update_value_data_.data()),
+          update_value_data_.size()) {
+}
+
+UpdateValue::UpdateValue(UpdateValue&& rhs)
+    : field_name_(rhs.field_name_)
+    , update_value_data_(rhs.update_value_data_)
+    , update_value_view_(
+          (rhs.update_value_view_.content() != nullptr &&
+                   rhs.update_value_view_.size() == 0 ?
+               (void*)"" :
+               update_value_data_.data()),
+          update_value_data_.size()) {
+}
+
+UpdateValue::~UpdateValue() {
+}
+
+UpdateValue& UpdateValue::operator=(const UpdateValue& rhs) {
+  if (this != &rhs) {
+    field_name_ = rhs.field_name_;
+    update_value_data_ = rhs.update_value_data_;
+    update_value_view_ = UntypedDatumView(
+        rhs.update_value_view_.content() != nullptr &&
+                rhs.update_value_view_.size() == 0 ?
+            (void*)"" :
+            update_value_data_.data(),
+        update_value_data_.size());
+  }
+
+  return *this;
+}
+
+UpdateValue& UpdateValue::operator=(UpdateValue&& rhs) {
+  field_name_ = std::move(rhs.field_name_);
+  update_value_data_ = std::move(rhs.update_value_data_);
+  update_value_view_ = UntypedDatumView(
+      rhs.update_value_view_.content() != nullptr &&
+              rhs.update_value_view_.size() == 0 ?
+          (void*)"" :
+          update_value_data_.data(),
+      update_value_data_.size());
+  return *this;
+}
+
+Status UpdateValue::check(const ArraySchema& array_schema) const {
```

Comment:
There's an ad-hoc type checking here which is not specific to this class. I don't want to broaden the scope for this PR too much, but whenever you get back to this, I think it would be worthwhile to pull the body of this function out to something like `check_compatible(ArraySchema&, std::string field_name, UntypedDatumView&)` (which would be called here if needed for convenience). Ideally in the future we'll also have a version taking a typed datum.